src/platform_impl/web/event_loop/runner.rs

```diff
@@ -348,16 +347,11 @@ impl<T: 'static> Shared<T> {
                 &mut control,
             );
 
-            // Then we resize the canvas to the new size and send a `Resized` event:
+            // Then we resize the canvas to the new size, a new
+            // `Resized` event will be sent by the `ResizeObserver`:
             if current_size != new_size {
-                backend::set_canvas_size(&canvas, crate::dpi::Size::Physical(new_size));
-                self.handle_single_event_sync(
-                    Event::WindowEvent {
-                        window_id: id,
-                        event: crate::event::WindowEvent::Resized(new_size),
-                    },
-                    &mut control,
-                );
+                let canvas = &canvas.borrow();
+                backend::set_canvas_size(canvas.window(), canvas.raw(), new_size.cast());
```

Comment:
Currently this is the intention of `ScaleFactorChanged::new_inner_size`. But I did address this issue in #2860, if users don't want Winit to change or overwrite a size set by CSS, they have to change `new_inner_size` to the old value.

I agree that this is not desirable on Web, but I'm not sure this can be changed with the current API without introducing inconsistent behavior between backends. There might be room for some design work to allow for this, I was thinking of changing `new_inner_size` from `&mut PhysicalSize<u32>` to `&mut Option<PhysicalSize<u32>>`. In any case, this is off-topic for this PR.

An alternative we could apply with how things currently work in Winit, is to set the `new_inner_size` suggested by `ScaleFactorChanged` to one that doesn't actually change anything, ergo doesn't scale. The argument would be that `new_inner_size` is "suggested by the OS", we could argue that Web makes the suggestion to not change the size.

Would like some input from other maintainers about this topic.

Reply:
Oops, I completely forgot that `new_inner_size` existed. The current situation of requiring that you manually set it to the old value indeed isn't ideal, though. 

The ideal solution in my opinion would be to find out what the new size is as determined by CSS, report that in `new_inner_size`, and then only set `width`/`height` if it was changed. That's what I implemented in #2074.

> An alternative we could apply with how things currently work in Winit, is to set the `new_inner_size` suggested by `ScaleFactorChanged` to one that doesn't actually change anything, ergo doesn't scale. The argument would be that `new_inner_size` is "suggested by the OS", we could argue that Web makes the suggestion to not change the size.

That's basically what I'm suggesting, except that it needs be the value that doesn't change the _new_ size of the canvas, not the size before the scale factor changed.

Reply:
Tracked in https://github.com/rust-windowing/winit/issues/2864.